test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/BatchRequestTests/Tests/BatchRequestClientTests.cs

```diff
@@ -0,0 +1,203 @@
+        {
+            services.ConfigureControllers(typeof(BatchRequestTestsController), typeof(MetadataController));
+
+            services.AddControllers().AddOData(opt =>
+            {
+                opt.EnableQueryFeatures().AddRouteComponents("odata", DefaultEdmModel.GetEdmModel(), new DefaultODataBatchHandler());
+                opt.RouteOptions.EnableNonParenthesisForEmptyParameterFunction = true;
+            });
+        }
+    }
+
+    public BatchRequestClientTests(TestWebApplicationFactory<TestsStartup> fixture) : base(fixture)
+    {
+        if (Client.BaseAddress == null)
+        {
+            throw new ArgumentNullException(nameof(Client.BaseAddress), "Base address cannot be null");
+        }
+
+        _baseUri = new Uri(Client.BaseAddress, "odata/");
+
+        _context = new Container(_baseUri)
+        {
+            HttpClientFactory = HttpClientFactory
+        };
+
+        ResetDefaultDataSource();
+    }
+
+    [Fact]
+    public async Task AddAndModifiedBatchRequestsTest()
+    {
+        // Arrange
+
+        // POST Requests
+        var account = new Account
+        {
+            AccountID = 110,
+            CountryRegion = "US"
+        };
+        _context.AddToAccounts(account);
+
+        var paymentInstrument = new PaymentInstrument
+        {
+            PaymentInstrumentID = 102910,
+            FriendlyName = "102 batch new PI",
+            CreatedDate = new DateTimeOffset(new DateTime(2013, 12, 29, 11, 11, 57))
+        };
+        _context.AddRelatedObject(account, "MyPaymentInstruments", paymentInstrument);
+
+        var billingStatement = new Statement
+        {
+            StatementID = 102910010,
+            TransactionDescription = "Digital goods: PC",
+            Amount = 1000
+        };
+        _context.AddRelatedObject(paymentInstrument, "BillingStatements", billingStatement);
+
+        // PATCH Request
+        var accountToUpdate = _context.Accounts.Where(a => a.AccountID == 107).Single();
+        Assert.NotNull(accountToUpdate);
+        var now = DateTimeOffset.Now;
+        accountToUpdate.UpdatedTime = now;
+        accountToUpdate.AccountInfo = new AccountInfo
+        {
+            FirstName = "John",
+            LastName = "Doe"
+        };
+        _context.UpdateObject(accountToUpdate);
+
+        // Act
+        var response = await _context.SaveChangesAsync(SaveChangesOptions.BatchWithSingleChangeset | SaveChangesOptions.UseRelativeUri);
+
+        // Assert
+        Assert.Equal(4, response.Count());
+
+        var changeResponses = response.Cast<ChangeOperationResponse>();
+
+        Assert.All(changeResponses, changeResponse => Assert.True(changeResponse.StatusCode == 200 || changeResponse.StatusCode == 201));
+
+        var accountCreatedResponse = changeResponses.FirstOrDefault(r => r.StatusCode == 201 && r.Descriptor is EntityDescriptor descriptor && descriptor.Entity is Account);

```

Comment:
Retrieve the request by status code seems brittle and counter-intuitive. Does the `SaveChangesAsync` method make any guarantees in terms of the order of responses? If so we can use that and also implicitly test that behaviour. If not, can we assert based on the type of request?